src/main/java/org/apache/sysml/hops/AggUnaryOp.java

```diff
@@ -627,28 +632,68 @@ private boolean isUnaryAggregateOuterCPRewriteApplicable()
 	private Lop constructLopsTernaryAggregateRewrite(ExecType et) 
 		throws HopsException, LopsException
 	{
-		Hop input1 = getInput().get(0);
+		BinaryOp input1 = (BinaryOp)getInput().get(0);
 		Hop input11 = input1.getInput().get(0);
 		Hop input12 = input1.getInput().get(1);
 		
-		Lop in1 = null;
-		Lop in2 = null;
-		Lop in3 = null;
-		
-		if( input11 instanceof BinaryOp && ((BinaryOp)input11).getOp()==OpOp2.MULT )
-		{
-			in1 = input11.getInput().get(0).constructLops();
-			in2 = input11.getInput().get(1).constructLops();
-			in3 = input12.constructLops();
-		}
-		else if( input12 instanceof BinaryOp && ((BinaryOp)input12).getOp()==OpOp2.MULT )
-		{
-			in1 = input11.constructLops();
-			in2 = input12.getInput().get(0).constructLops();
-			in3 = input12.getInput().get(1).constructLops();
+		Lop in1 = null, in2 = null, in3 = null;
+		boolean handled = false;
+
+		if (input1.getOp() == OpOp2.POW) {
+			switch ((int)((LiteralOp)input12).getLongValue()) {
+			case 3:
+				in1 = input11.constructLops();
+				in2 = in1;
+				in3 = in1;
+				break;
+			default:
+				throw new AssertionError("unreachable; only applies to power 3");
+			}
+			handled = true;
+		} else if (input11 instanceof BinaryOp ) {
+			BinaryOp b11 = (BinaryOp)input11;
+			switch (b11.getOp()) {
+			case MULT: // A*B*C case
+				in1 = input11.getInput().get(0).constructLops();
+				in2 = input11.getInput().get(1).constructLops();
+				in3 = input12.constructLops();
+				handled = true;
+				break;
+			case POW: // A*A*B case
+				Hop b112 = b11.getInput().get(1);
+				if ( !(input12 instanceof BinaryOp && ((BinaryOp)input12).getOp()==OpOp2.MULT)
```

Comment:
just for simplification, please use the existing HopRewriteUtil method for these checks.